### PR TITLE
Add opam pin remove --all

### PR DESCRIPTION
**Pull request by kit-ty-kate**
I often find the need to clean all pins from the current switch i am working on, be it because the pin packages have all been released or other reasons, and listing all the packages or repos one by one can be extremely tedious.

### CODE REVIEW

**Comment by avsm** [comment]
Yes please!  I also have the case that it's difficult to undo an `opam install`'s pins

```
opam install .
# proceeds to pin all the local packages
# how do i undo the previous command? opam remove doesn't unpin them.
```

**Comment by kit-ty-kate** [comment]
Could you open a new issue to track that problem?

File: src/client/opamCommands.ml [hunk 1]
```diff
@@ -3374,6 +3380,12 @@ let pin ?(unpin_only=false) cli =
       else
         (OpamSwitchState.drop @@ OpamClient.PIN.unpin st ~action to_unpin;
          `Ok ())
+    | `remove_all ->
+      OpamGlobalState.with_ `Lock_none @@ fun gt ->
+      OpamSwitchState.with_ `Lock_write gt @@ fun st ->
+      let to_unpin = OpamPackage.Set.to_list_map OpamPackage.name (OpamPinned.packages st) in
+      OpamSwitchState.drop (OpamClient.PIN.unpin st ~action to_unpin);
```
Comment (rjbou) [inline src/client/opamCommands.ml:3387]:
```suggestion
      OpamSwitchState.drop @@ OpamClient.PIN.unpin st ~action to_unpin;
```
Just to keep the same style

File: src/client/opamCommands.ml [hunk 2]
```diff
@@ -3280,6 +3284,8 @@ let pin ?(unpin_only=false) cli =
         `list
       | Some `scan, [url] ->
         `scan url
+      | Some `remove, _ when all ->
+        `remove_all
```
Comment (rjbou) [inline src/client/opamCommands.ml:3288]: We should error when remove is given with an argument and `--all`.

File: src/client/opamCommands.ml [hunk 3]
```diff
@@ -3374,6 +3380,12 @@ let pin ?(unpin_only=false) cli =
       else
         (OpamSwitchState.drop @@ OpamClient.PIN.unpin st ~action to_unpin;
          `Ok ())
+    | `remove_all ->
+      OpamGlobalState.with_ `Lock_none @@ fun gt ->
+      OpamSwitchState.with_ `Lock_write gt @@ fun st ->
+      let to_unpin = OpamPackage.Set.to_list_map OpamPackage.name (OpamPinned.packages st) in
```
Comment (rjbou) [inline src/client/opamCommands.ml:3386]:
```suggestion
      let to_unpin =
        OpamPackage.Set.to_list_map OpamPackage.name (OpamPinned.packages st)
      in
```

File: tests/reftests/pin.test [hunk 4]
```diff
@@ -175,3 +175,28 @@ echo "another-inexistant" "inexistant"
...
+  - remove nip dev
+  - remove qux dev
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   bar.dev
+-> removed   foo.1
+-> removed   nip.dev
+-> removed   qux.dev
+Done.
+### opam pin
```
Comment (rjbou) [inline tests/reftests/pin.test:202]: we should add also a test with `--no-action --all`

File: tests/reftests/pin.test [hunk 5]
```diff
@@ -175,3 +175,28 @@ echo "another-inexistant" "inexistant"
 -> installed qux.dev
 -> installed bar.dev
 Done.
+### opam pin
+bar.dev    rsync             file://${BASEDIR}/bar
+foo.1      local definition
+nip.dev    rsync             file://${BASEDIR}/nip
+qux.dev    rsync             file://${BASEDIR}/qux
+### opam pin remove -a
```
Comment (rjbou) [inline tests/reftests/pin.test:183]:
```suggestion
### opam pin remove --all
```
better always explicit argument in tests

File: src/core/opamStd.mli [hunk 6]
```diff
@@ -29,6 +29,7 @@ module type SET = sig
   val choose_opt: t -> elt option
 
   val of_list: elt list -> t
+  val to_list_map: (elt -> 'b) -> t -> 'b list
```
Comment (rjbou) [inline src/core/opamStd.mli:32]: Can these changes be in a separate commit?

**Comment by kit-ty-kate** [comment]
All done

File: src/client/opamCommands.ml [hunk 7]
```diff
@@ -3147,6 +3147,10 @@ let pin ?(unpin_only=false) cli =
        changes, and may also be used to keep a package that was removed \
        upstream."
   in
+  let all =
+    mk_flag ~cli (cli_from cli2_2) ["a"; "all"]
+      "When unpinning, removes all pin in the given switch."
```
Comment (dra27) [inline src/client/opamCommands.ml:3152]:
```suggestion
      "When unpinning, removes all pins in the given switch."
```

**Review by dra27 — COMMENTED** [review]
Tiny tweak to the message.

More importantly, `-a` has a lot of different meanings across opam already, so this doesn't technically make things that much worse, however `opam remove -a` removes automatically-installed packages but `opam pin remove -a` would now mean remove _all_ pins. Might it be worth either using a different letter (`opam admin` already has `-A` for a `--all` option) or just not having a short form at all for this one?

File: src/client/opamCommands.ml [hunk 8]
```diff
@@ -3280,6 +3284,11 @@ let pin ?(unpin_only=false) cli =
         `list
       | Some `scan, [url] ->
         `scan url
+      | Some `remove, params when all ->
+        if params = [] then
+          `remove_all
+        else
+          `incorrect
```
Comment (rjbou) [inline src/client/opamCommands.ml:3291]:
```suggestion
      | Some `remove, [] when all ->
        `incorrect
      | Some `remove, params when all ->
        `remove_all
```

**Comment by avsm** [comment]
I'm not convinced we need this PR because:

1) `opam pin remove -a` has a different meaning from `opam remove -a` (as @dra27 notes)
2) `opam pin remove $(opam pin -s)` accomplishes what this PR does -- we could document this.
3) `opam pin install` and `opam pin remove` are not currently symmetric, and there's no easy way to do this.

I'd recommend not adding more CLI options to opam when there are existing unix-style solutions that work, and focussing on fixing the really rough edges like 3) above which is hard to do (as you cannot easily probe the set of locally pinned packages without some wildcard magic)

File: tests/reftests/pin.test [hunk 9]
```diff
@@ -176,13 +176,66 @@ echo "another-inexistant" "inexistant"
...
 foo.1      local definition
 nip.dev    rsync             file://${BASEDIR}/nip
 qux.dev    rsync             file://${BASEDIR}/qux
-### opam pin remove --no-action bar.dev
+### OPAMNODEPEXTS=1
+### opam pin
+bar.dev    rsync             file://${BASEDIR}/bar
+foo.1      local definition
+nip.dev    rsync             file://${BASEDIR}/nip
+qux.dev    rsync             file://${BASEDIR}/qux
```
Comment (rjbou) [inline tests/reftests/pin.test:198]:
```suggestion
```
the opam pin is doubled

**Review by rjbou — APPROVED** [review]

File: src/client/opamCommands.ml [hunk 10]
```diff
@@ -3184,6 +3184,10 @@ let pin ?(unpin_only=false) cli =
        changes, and may also be used to keep a package that was removed \
        upstream."
   in
+  let all =
+    mk_flag ~cli (cli_from cli2_2) ["all"]
+      "When unpinning, removes all pins in the given switch."
```
Comment (rjbou) [inline src/client/opamCommands.ml:3189]: Should we add a word about this command being a no-op on other pin commands ?

Comment (kit-ty-kate) [reply src/client/opamCommands.ml:3189]: I chose to make the use of `--all` outside of `remove` without any parameter an error instead. Does that work for you?

File: tests/reftests/pin.test (hunk 9 above)
Comment (kit-ty-kate) [reply tests/reftests/pin.test]: Done

File: src/client/opamCommands.ml (hunk 10 above)
Comment (rjbou) [reply src/client/opamCommands.ml:3189]: good for me!

File: tests/reftests/pin.test [hunk 11]
```diff
@@ -176,13 +176,66 @@ echo "another-inexistant" "inexistant"
 -> installed bar.dev
 Done.
 ### : Test opam pin remove <pkg>.<version>
+### opam pin remove --no-action bar.dev
```
[the remark this answers is not in the document]
Comment (rjbou) [reply tests/reftests/pin.test:179]: yes, i had in mind to put it later indeed. Perfect!

**Review by rjbou — APPROVED** [review]
